File: src/world/shipparts/missileLauncher.lua
local Timer = require("timer")

local MissileLauncher = class()

function MissileLauncher:__create()
	self.charged = true
	self.rechargeTimer = Timer(1)

	return self
end

function MissileLauncher.process(orders)
	local shoot = false
	for _, order in ipairs(orders) do
		if order == "shoot" then shoot = true end
	end
	return shoot
end

function MissileLauncher:update(inputs, location)
	if not self.charged then
		if self.rechargeTimer:ready(inputs.dt) then
			self.charged = true
		end
	else
		if inputs.controls.gun then
			-- Check if there is a part one block infront of the gun.
			local getPart = inputs.getPart
			if not getPart(location, {0, 1}) then
				self.charged = false
				-- Spawn Shot
				return {"missile", {0, 1, 1}, 1}
			end
		end
	end
end

return MissileLauncher

File: src/world/shipparts/missileLauncher.rs
use mlua::prelude::{LuaFunction, LuaNumber, LuaResult, LuaTable, LuaValue};
use mlua::{Lua, Result, ToLua, UserData, UserDataFields, UserDataMethods};
use crate::world::types::{Location, Module, ModuleInputs, WorldEvent};
use crate::timer::{Timer};

struct MissileLauncher {
    charged: bool,
    rechargeTimer: Timer,
}

impl MissileLauncher {
    fn process(&self, orders: Vec<String>) -> bool {
        // TODO: functionalize
        let shoot: bool;
        for order in orders {
            if order == "shoot" {
                shoot = true;
            }
        }

        shoot
    }
}

impl Module for MissileLauncher {
    fn update(&mut self, inputs: ModuleInputs, location: Location) -> Option<WorldEvent> {
        if !self.charged {
            if self.rechargeTimer.ready(inputs.dt) {
                self.charged = true;
            }
            None
        } else {
            if inputs.controls.gun {
                // Check if there is a part one block in front of the gun.
                if !inputs.getPart(location, (0, 1)) {
                    self.charged = false;
                    ("missile", (0, 1, 1), 1)
                }
            }
        }
    }
}

impl UserData for MissileLauncher {
    fn add_fields<'lua, F: UserDataFields<'lua, Self>>(fields: &mut F) {
        fields.add_field_method_get("limit", |_, this| Ok(this.limit));
        fields.add_field_method_get("time", |_, this| Ok(this.time));
    }

    fn add_methods<'lua, M: UserDataMethods<'lua, Self>>(methods: &mut M) {
        methods.add_method_mut("ready", |_, this, dt: f64| Ok(Timer::ready(this, dt)));
    }
}

pub fn lua_module(lua: &Lua) -> Result<LuaTable> {
    let exports = lua.create_table()?;

    let metatable = lua.create_table()?;
    exports.set_metatable(Some(metatable));

    metatable.set(
        "__call",
        lua.create_function(|lua| {
            MissileLauncher {
                charged: true,
                rechargeTimer: Timer {limit: 1, time: 1},
            }
            .to_lua(lua)
        })?,
    )?;

    Ok(exports)
}
